fe1-web/src/features/popcha/screens/__tests__/PopchaScanner.test.tsx

```diff
@@ -0,0 +1,214 @@
+      const url = new URL(mockUrl.toString());
+      url.searchParams.delete('response_type');
+      await testInvalidUrl(url.toString());
+    });
+
+    it('url without scope shows error message', async () => {
+      const url = new URL(mockUrl.toString());
+      url.searchParams.delete('scope');
+      await testInvalidUrl(url.toString());
+    });
+
+    it('url with invalid response_type shows error message', async () => {
+      const url = new URL(mockUrl.toString());
+      url.searchParams.set('response_type', 'scope invalid response type');
+      await testInvalidUrl(url.toString());
+    });
+
+    it('url with invalid response_mode shows error message', async () => {
+      const url = new URL(mockUrl.toString());
+      url.searchParams.set('response_mode', 'invalid response mode');
+      await testInvalidUrl(url.toString());
+    });
+
+    it('login_hint does not match current laoId shows error message', async () => {
+      const url = new URL(mockUrl.toString());
+      url.searchParams.set('login_hint', 'invalid login hint');
+      await testInvalidUrl(url.toString());
+    });
+  });
+
+  describe('valid url sends correct response', () => {
+    it('valid url sends correct response', async () => {
+      const { getByTestId } = render(
+        <FeatureContext.Provider value={contextValue}>
+          <PopchaScanner />
+        </FeatureContext.Provider>,
+      );
+
+      const scannerButton = getByTestId('popcha_scanner_button');
+      fireEvent.press(scannerButton);
+      fireScan(mockUrl.toString());
+      await waitFor(() => expect(mockToastShow).toHaveBeenCalledTimes(0));
+    });
+
+    it('valid response mode does not show error message', async () => {
+      const url = new URL(mockUrl.toString());
+      url.searchParams.set('response_mode', 'query');
+      const { getByTestId } = render(
+        <FeatureContext.Provider value={contextValue}>
+          <PopchaScanner />
+        </FeatureContext.Provider>,
+      );
+
+      const scannerButton = getByTestId('popcha_scanner_button');
+      fireEvent.press(scannerButton);
+      fireScan(url.toString());
+      await waitFor(() => expect(mockToastShow).toHaveBeenCalledTimes(0));
+    });
+  });
+
+  // TODO: fix these tests
+
+  describe('correct behavior when sending request', () => {
+    it('successful request closes scanner', async () => {
+      const url = new URL(mockUrl.toString());
+      (sendPopchaAuthRequest as jest.Mock).mockReturnValue(Promise.resolve());
+      const { getByTestId, toJSON } = render(
+        <FeatureContext.Provider value={contextValue}>
+          <PopchaScanner />
+        </FeatureContext.Provider>,
+      );
+      const scannerButton = getByTestId('popcha_scanner_button');
+      fireEvent.press(scannerButton);
+      fireScan(url.toString());
+      await waitFor(() => {
+        expect(mockToastShow).toHaveBeenCalledTimes(0);
+        expect(toJSON()).toMatchSnapshot();
+      });
+    });
```

Comment:
This test currently does not work as planned since in the snapshot it seems the scanner is still open, I am not sure why though

Reply:
Are you sure to close the scanner correctly?
Maybe you could try to mock the QR code closing function and then check that it has been called.

It's quite hard for me to understand what is not working from just reading the code.

Reply:
The problem was that in the snapshot file, I could see that the scanner didn't close, but it seems that the problem was I was not waiting long enough for. So I modified in the following way:

```Typescript
      await waitFor(() => {
        // wait for scanner to close
        expect(getByText(STRINGS.popcha_open_scanner)).toBeTruthy();
        expect(mockToastShow).toHaveBeenCalledTimes(0);
        expect(toJSON()).toMatchSnapshot();
      });
```